### PR TITLE
TIP-1159 [New calculator]

### PR DESCRIPTION
**Pull request by pierallard**
This PR uses a new way to calculate the completeness using mask.

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php [hunk 1]
```diff
@@ -1,131 +1,100 @@
...
- * Calculates the completenesses for a provided product.
- *
- * @author    Damien Carcel (damien.carcel@akeneo.com)
- * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
- * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ * @author    Pierre Allard <pierre.allard@akeneo.com>
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
 class CompletenessCalculator implements CompletenessCalculatorInterface
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: We could rename it ProductCompletenessCalculator without interface, missing spec

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php [hunk 2]
```diff
@@ -1,131 +1,100 @@
...
-        foreach ($family->getAttributeRequirements() as $attributeRequirement) {
-            $channel = $attributeRequirement->getChannel();
-            if (!$channels->contains($channel)) {
-                $channels->add($channel);
-            }
-        }
+        return $this->getCompletenessFamilyMasks->fromFamilyCodes($familyCodes);
+    }
+
+    private function applyMask(
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: this logic should be in the family mask in my opinion

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php [hunk 3]
```diff
@@ -1,131 +1,100 @@
...
+    ): ProductCompletenessCollection {
+        $productMask = $completenessProductMask->mask();
+
+        return new ProductCompletenessCollection($completenessProductMask->getId(), array_map(
+            function (CompletenessFamilyMaskPerChannelAndLocale $familyMaskPerChannelAndLocale) use ($productMask) {
+                $diff = array_diff($familyMaskPerChannelAndLocale->mask(), $productMask);
 
-        return $channels;
+                $missingAttributeCodes = array_map(function (string $mask) {
+                    return substr($mask, 0, strpos($mask, '-'));
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: can we put a constant for '-' like SEPARATOR and a phpdoc to indicate that this separator should not be wllowed in attribute code (which is the case, I checked).

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php [hunk 4]
```diff
@@ -1,131 +1,100 @@
...
+        return $this->getCompletenessFamilyMasks->fromFamilyCodes($familyCodes);
+    }
+
+    private function applyMask(
+        CompletenessFamilyMask $familyMask,
+        CompletenessProductMask $completenessProductMask
+    ): ProductCompletenessCollection {
+        $productMask = $completenessProductMask->mask();
+
+        return new ProductCompletenessCollection($completenessProductMask->getId(), array_map(
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: avoid inline please

File: src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php (hunk 1 above)
Comment (anaelChardan) [reply src/Akeneo/Pim/Enrichment/Component/Product/Completeness/CompletenessCalculator.php]: Maybe it's another PR

**Pull request merged**